pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,599 @@
+	return getPath(b.podUID, b.volNamespace, b.volName, b.pvName, b.plugin.host)
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDown() error {
+	if len(b.volNamespace) == 0 || len(b.volName) == 0 {
+		return nil
+	}
+	// Unmount from pod
+	mountPath := b.GetPath()
+
+	err := b.TearDownAt(mountPath)
+	if err != nil {
+		glog.Errorf("storageos: Unmount from pod failed: %v", err)
+		return err
+	}
+
+	// Find device name from global mount
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.volNamespace, b.volName)
+	devicePath, _, err := mount.GetDeviceNameFromMount(b.mounter, globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed when getting device from global mount: %v", err)
+		return err
+	}
+
+	// Unmount from plugin's disk global mount dir.
+	err = b.TearDownAt(globalPDPath)
+	if err != nil {
+		glog.Errorf("Detach failed during unmount: %v", err)
+		return err
+	}
+
+	// Detach loop device
+	err = b.manager.DetachVolume(b, devicePath)
+	if err != nil {
+		glog.Errorf("Detach device %s failed for volume %s: %v", devicePath, b.pvName, err)
+		return err
+	}
+
+	glog.V(4).Infof("Successfully unmounted StorageOS volume %s and detached devices", b.pvName)
+
+	return nil
+}
+
+// Unmounts the bind mount, and detaches the disk only if the PD
+// resource was the last reference to that disk on the kubelet.
+func (b *storageosUnmounter) TearDownAt(dir string) error {
+	if err := util.UnmountPath(dir, b.mounter); err != nil {
+		glog.V(4).Infof("Unmounted StorageOS volume %s failed with: %v", b.pvName, err)
+	}
+	if err := b.manager.UnmountVolume(b); err != nil {
+		glog.V(4).Infof("Mount reference for volume %s could not be removed from StorageOS: %v", b.pvName, err)
+	}
+	return nil
+}
+
+type storageosDeleter struct {
+	*storageos
+	apiAddr    string
+	apiUser    string
+	apiPass    string
+	apiVersion string
+}
+
+var _ volume.Deleter = &storageosDeleter{}
+
+func (d *storageosDeleter) GetPath() string {
+	return getPath(d.podUID, d.volNamespace, d.volName, d.pvName, d.plugin.host)
+}
+
+func (d *storageosDeleter) Delete() error {
+	return d.manager.DeleteVolume(d)
+}
+
+type storageosProvisioner struct {
+	*storageos
+	options volume.VolumeOptions
+}
+
+var _ volume.Provisioner = &storageosProvisioner{}
+
+func (c *storageosProvisioner) Provision() (*v1.PersistentVolume, error) {
+
+	// Apply ProvisionerParameters (case-insensitive). We leave validation of
+	// the values to the cloud provider.
+	for k, v := range c.options.Parameters {
+		switch strings.ToLower(k) {
+		case "description":
+			c.description = v
+		case "pool":
+			c.pool = v
+		case "fstype":
+			c.fsType = v
+		default:
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+		}
+	}
+
+	// Set from PVC
+	c.volName = c.options.PVName
+	c.volNamespace = c.options.PVC.Namespace
```

Comment:
is this right? I thought volNamespace was the storageOS namespace, not the k8s namespace